source/generators/MuonAngleGenerator.cc

```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
-  zenith  *= pi;
-  azimuth *= pi;
-
-  dir.setX(sin(zenith) * sin(azimuth));
-  dir.setY(-cos(zenith));
-  dir.setZ(-sin(zenith) * cos(azimuth));
-
-  dir *= *rPhi_;
+
+  // Bool to check if zenith has a valid value. If not then resample
+  G4bool invalid_evt = true;
+
+  while(invalid_evt){
+  
+    // Amount to smear the randomly sampled zenith/azimuth values by
+    G4double zen_BW_smear{1.0e6}; 
+    G4double az_BW_smear{1.0e6};
+
+    // Generate random index weighted by the bin contents
+    G4int RN_indx = discr_dist_(RN_engine_);
+
+    // Loop over the zenith values and find the corresponding bin width to smear
+    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
+        
+        // Catch very last bin
+        if (zen_BW_smear == 1e6 && i == zenith_bins_.size()-2){
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+              && zeniths_[RN_indx] <= zenith_bins_[i+1]){
+            
+              zen_BW_smear = zen_BW_[i];
+            }
+        }
+        else {
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+                && zeniths_[RN_indx] < zenith_bins_[i+1]){
+              
+              zen_BW_smear = zen_BW_[i];
+
+          }
+        }
+    
+    }
+
+    // Loop over the azimuth values and find the corresponding bin width to smear
+    for (G4int i = 0; i < azimuth_bins_.size()-1; i++){
+        
+        // Include last bin in check
+        if (az_BW_smear == 1e6 && i == azimuth_bins_.size()-2){
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+              && azimuths_[RN_indx] <= azimuth_bins_[i+1]){
+            
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+        else {
+
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+                && azimuths_[RN_indx] < azimuth_bins_[i+1]){
+              
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+
+    }
+
+    // Check if the smear values are set properly
+    if (az_BW_smear == 1.0e6 || zen_BW_smear == 1.0e6 )
+      G4Exception("[MuonAngleGenerator]", "GetDirection()",
+                FatalException, " bin smear value was not set correctly ");
+
+    // Gaussian dist to smear by bin widths in azimuth and zenith
+    std::normal_distribution<G4double> Gauss_az(0, az_BW_smear);
+    std::normal_distribution<G4double> Gauss_zen(0, zen_BW_smear);
+    
+    // Get the Gaussian smear values 
+    G4double az_smear  = Gauss_az(RN_engine_az_);
+    G4double zen_smear = Gauss_zen(RN_engine_zen_);
+
+    // Correct sampled values by smearing
+    G4double az_samp  = azimuths_[RN_indx] + az_smear;
+    G4double zen_samp = zeniths_[RN_indx]  + zen_smear;
+
+    // Catch negative value and resample if so
+    if (zen_samp < 0.0)
+        invalid_evt = true;
+    else
+        invalid_evt = false;
+
+    // Calculate the vector components of the muon
+    dir.setX(sin(zen_samp) * sin(az_samp));
+    dir.setY(-cos(zen_samp));
+    dir.setZ(-sin(zen_samp) * cos(az_samp));
```

Comment:
This assumes that azimuth is the angle between muon direction and -Z axis. Is it the intented behaviour? I would naively consider the azimuth to be between muon direction and +X. 
Also, which is the value of `azimuth_rotation` for NEXT100? I see that `NEXT100_muons_lsc` and `NEXT100_muons_hallA` use the value of 150 degrees. The rotation here is performed anticlockwise throughout Y axis (meaning -150). Is it the correct orientation?

Reply:
All these slightly non-intuitive definitions are due to the difference in axis orientation between NEXT, set so that positive z points towards the cathode, and the 'natural' orientation where positive z points upwards. Additionally, the measurements of the flux in the lab are oriented so that the azimuthal angle 0 is at north so we need to have the azimuth angle around Y (which points up in next) and rotate the axes in the input info to the orientation in NEXT. Since NEXT100 will be in the same shielding as NEW (unless things have changed since I was in NEXT) we always assumed the 150 degrees rotation of the axes would be the same but it's a configuration variable in case of changes there.

There's PR, that's probably very out of date, which updates a lot of important things for muon generation (#69 , think) to make it closer to reality and to aid normalisation, maybe worthwhile having a look.

Hope that helps.

Reply:
Thanks for the clarification!